internal/controller/artifactgenerator_drift.go

```diff
@@ -0,0 +1,138 @@
+//   - "ArtifactMissing" - artifact is missing from storage
+//   - "ArtifactCorrupted" - artifact exists in storage but fails integrity verification
+//   - "ExternalArtifactsNotFound" - failed to query in-cluster external artifacts
+//   - "ExternalArtifactsChanged" - in-cluster external artifacts differ from inventory
+//   - "NoDriftDetected" - no drift detected and the storage is up to date
+func (r *ArtifactGeneratorReconciler) detectDrift(ctx context.Context,
+	obj *swapi.ArtifactGenerator,
+	currentSourcesDigest string) (bool, string) {
+	// Setup logger on debug level.
+	log := ctrl.LoggerFrom(ctx).V(1)
+
+	if conditions.IsFalse(obj, meta.ReadyCondition) {
+		log.Info("Drift detected, previous reconciliation failed")
+		return true, "NotReady"
+	}
+
+	if obj.GetGeneration() != conditions.GetObservedGeneration(obj, meta.ReadyCondition) {
+		log.Info("Drift detected, generation has changed",
+			"old", conditions.GetObservedGeneration(obj, meta.ReadyCondition),
+			"new", obj.GetGeneration())
+		return true, "GenerationChanged"
+	}
+
+	if obj.Status.ObservedSourcesDigest != currentSourcesDigest {
+		log.Info("Drift detected, sources have changed",
+			"old", obj.Status.ObservedSourcesDigest,
+			"new", currentSourcesDigest)
+		return true, "SourcesChanged"
+	}
+
+	if len(obj.Status.Inventory) != len(obj.Spec.OutputArtifacts) {
+		log.Info("Drift detected, number of output artifacts has changed",
+			"old", len(obj.Status.Inventory),
+			"new", len(obj.Spec.OutputArtifacts))
+		return true, "ArtifactsChanged"
+	}
+
+	for _, eaRef := range obj.Status.Inventory {
+		storagePath := storage.ArtifactPath(sourcev1.ExternalArtifactKind, eaRef.Namespace, eaRef.Name, eaRef.Filename)
+		artifact := meta.Artifact{
+			Digest: eaRef.Digest,
+			Path:   storagePath,
+		}
+		if !r.Storage.ArtifactExist(artifact) {
+			log.Info("Drift detected, artifact missing from storage",
+				"artifact", fmt.Sprintf("%s/%s/%s", sourcev1.ExternalArtifactKind, eaRef.Namespace, eaRef.Name),
+				"path", storagePath)
+			return true, "ArtifactMissing"
+		}
+		if err := r.Storage.VerifyArtifact(artifact); err != nil {
+			log.Error(err, "Artifact integrity verification failed, deleting corrupted artifact",
+				"artifact", fmt.Sprintf("%s/%s/%s", sourcev1.ExternalArtifactKind, eaRef.Namespace, eaRef.Name))
+			if err := r.Storage.Remove(artifact); err != nil {
+				log.Error(err, "Failed to remove corrupted artifact from storage",
+					"artifact", fmt.Sprintf("%s/%s/%s", sourcev1.ExternalArtifactKind, eaRef.Namespace, eaRef.Name))
+			}
+			return true, "ArtifactCorrupted"
+		}
+	}
+
+	eaDrift, err := r.detectExternalArtifactDrift(ctx, obj)
+	if err != nil {
+		log.Error(err, "Failed to verify in-cluster external artifacts for drift")
+		return true, "ExternalArtifactsNotFound"
+	}
+	if eaDrift {
+		log.Info("Drift detected, in-cluster external artifacts have changed")
+		return true, "ExternalArtifactsChanged"
+	}
```

Comment:
Since at this point of the drift detection we already know that the referenced sources are in-sync with the built artifacts in the storage, maybe we could optimize the reconciliation to avoid the fetching and building and reconciling only the in-cluster EA objects. I'd say this is not required for an MVP of the controller, so we can leave this for another PR